### PR TITLE
refactor(l1): snap sync code reorganization and error handling consolidation

### PR DESCRIPTION
**Pull request by pablodeymo**
## Summary

Major refactoring of the snap sync implementation (~6,500 lines across 7 files) to improve code organization, maintainability, and error handling. This PR completes all 5 phases of the snap sync refactoring plan.

## Motivation

The snap sync codebase had grown organically with:
- Large monolithic files (sync.rs at 1,648 lines, peer_handler.rs at 2,074 lines)
- Scattered error types across multiple modules
- Mixed concerns (server/client code, protocol messages, sync orchestration)
- Inconsistent naming conventions in healing modules

This refactoring addresses these issues through a structured, phased approach.

### Phase 1: Foundation ✅
- Created `snap/` module directory structure
- Moved server code from `snap.rs` to `snap/server.rs`
- Created `snap/constants.rs` with documented protocol constants
- Moved snap server tests to root test crate (`test/tests/p2p/snap_server_tests.rs`) to avoid `clippy::unwrap_used` lint denial inherited from the `ethrex-p2p` crate

### Phase 2: Protocol Layer ✅
- Split `rlpx/snap.rs` into `rlpx/snap/` directory:
  - `messages.rs` - Snap protocol message type definitions
  - `codec.rs` - RLPxMessage encoding/decoding implementations
  - `mod.rs` - Module re-exports

### Phase 3: Healing Unification ✅
- Created `sync/healing/` directory with unified structure:
  - `types.rs` - Shared healing types (HealingQueueEntry, StateHealingQueue)
  - `state.rs` - State trie healing (~420 lines)
  - `storage.rs` - Storage trie healing (~530 lines)
- Improved naming conventions:
  - `MembatchEntryValue` → `HealingQueueEntry`
  - `MembatchEntry` → `StorageHealingQueueEntry`
  - `Membatch` → `StorageHealingQueue`
  - `children_not_in_storage_count` → `missing_children_count`

### Phase 4: Sync Orchestration ✅
- Split `sync.rs` (1,648 lines) into focused modules:
  - `sync/full.rs` (~260 lines) - Full sync implementation
  - `sync/snap_sync.rs` (~1,100 lines) - Snap sync implementation
  - `sync/mod.rs` (~285 lines) - Syncer struct and orchestration
- Extracted snap client methods from `peer_handler.rs` to `snap/client.rs`:
  - `request_account_range`, `request_bytecodes`, `request_storage_ranges`
  - `request_state_trienodes`, `request_storage_trienodes`
  - Reduced `peer_handler.rs` from 2,060 to 670 lines (~68% reduction)

### Phase 5: Error Handling ✅
- Created unified `snap/error.rs` with `SnapError` enum:
  - `Store` - Storage layer errors
  - `Protocol` - Connection/protocol errors  
  - `Trie` - Trie operation errors
  - `RlpDecode` - RLP decoding errors
  - `PeerTable` - Peer table errors
  - `BadRequest` - Invalid request validation
  - `ValidationError` - Response validation failures
  - `NoTasks`, `NoAccountHashes`, `NoAccountStorages`, `NoStorageRoots`
  - `InternalError`, `FileSystem`, `SnapshotDir`, `TaskPanic`
  - `InvalidData`, `InvalidHash`
- Removed redundant error types: `SnapClientError`, `RequestStateTrieNodesError`
- Added `From<SnapError>` for `PeerConnectionError` for error propagation
- Kept `RequestStorageTrieNodesError` struct for request ID tracking

## Bug Fixes

Fixed several potential panics and edge cases in `snap/client.rs`:

- **Empty vector indexing**: Replaced direct `[0]` indexing with safe `.get(0).ok_or()` or `.first().unwrap_or()` to prevent panics on empty vectors
- **Zero chunk_size in `request_bytecodes`**: Added early return for empty input and adjusted `chunk_count` to not exceed input length
- **Zero chunk_size in `request_account_range`**: Ensured `chunk_count` doesn't exceed the hash range to avoid zero-sized chunks
- **Typo fix**: "Trie to get" → "Tried to get" in error message

## New Module Structure

```
crates/networking/p2p/
├── snap/
│   ├── mod.rs          # Re-exports
│   ├── server.rs       # Server-side request processing
│   ├── client.rs       # Client-side request methods (~1,439 lines)
│   ├── constants.rs    # Protocol constants
│   └── error.rs        # Unified SnapError type
├── rlpx/snap/
│   ├── mod.rs          # Re-exports
│   ├── messages.rs     # Message struct definitions
│   └── codec.rs        # RLPxMessage implementations
├── sync/
│   ├── mod.rs          # Syncer and orchestration (~285 lines)
│   ├── full.rs         # Full sync (~260 lines)
│   ├── snap_sync.rs    # Snap sync (~1,100 lines)
│   └── healing/
│       ├── mod.rs      # Re-exports
│       ├── types.rs    # Shared healing types
│       ├── state.rs    # State healing (~420 lines)
│       └── storage.rs  # Storage healing (~530 lines)
└── peer_handler.rs     # ETH protocol methods (~670 lines)

test/tests/p2p/
└── snap_server_tests.rs  # Snap server tests (moved from p2p crate)
```

## Testing

- ✅ `cargo check -p ethrex-p2p` - Compilation passes
- ✅ `cargo clippy -p ethrex-p2p` - No warnings
- ✅ `cargo test -p ethrex-p2p` - All 36 unit tests pass
- ✅ `cargo test -p ethrex-test snap_server` - All 12 snap server tests pass (moved to root test crate)

## Checklist

- [ ] Updated `STORE_SCHEMA_VERSION` (crates/storage/lib.rs) if the PR includes breaking changes to the `Store` requiring a re-sync.

---

## Refactoring Plan Details

<details>
<summary>Click to expand the full refactoring plan</summary>

### Overview

The Snap Sync implementation spans ~6,500 lines across 7 files. This plan provides a structured approach to simplify and improve the code.

#### Original Structure
| File | Lines | Purpose |
|------|-------|---------|
| `crates/networking/p2p/snap.rs` | 1,008 | Server-side request processing (90% tests) |
| `crates/networking/p2p/rlpx/snap.rs` | 389 | Protocol message definitions |
| `crates/networking/p2p/sync.rs` | 1,648 | Main sync orchestration |
| `crates/networking/p2p/sync/state_healing.rs` | 471 | State trie healing |
| `crates/networking/p2p/sync/storage_healing.rs` | 718 | Storage healing |
| `crates/networking/p2p/sync/code_collector.rs` | 102 | Bytecode collection |
| `crates/networking/p2p/peer_handler.rs` | 2,074 | Client-side snap requests (~800 lines snap-related) |

#### New Structure (After Phases 1-5)
| File | Purpose |
|------|---------|
| `crates/networking/p2p/snap/mod.rs` | Snap module re-exports |
| `crates/networking/p2p/snap/server.rs` | Server-side request processing |
| `crates/networking/p2p/snap/client.rs` | Client-side snap request methods (~1,439 lines) |
| `crates/networking/p2p/snap/constants.rs` | Centralized protocol constants |
| `crates/networking/p2p/rlpx/snap/mod.rs` | Protocol message re-exports |
| `crates/networking/p2p/rlpx/snap/messages.rs` | Message struct definitions |
| `crates/networking/p2p/rlpx/snap/codec.rs` | RLPxMessage implementations |
| `crates/networking/p2p/sync/mod.rs` | Sync orchestration (~285 lines) |
| `crates/networking/p2p/sync/full.rs` | Full sync implementation (~260 lines) |
| `crates/networking/p2p/sync/snap_sync.rs` | Snap sync implementation (~1,100 lines) |
| `crates/networking/p2p/sync/healing/mod.rs` | Healing module re-exports |
| `crates/networking/p2p/sync/healing/types.rs` | Shared healing types |
| `crates/networking/p2p/sync/healing/state.rs` | State healing (~420 lines) |
| `crates/networking/p2p/sync/healing/storage.rs` | Storage healing (~530 lines) |
| `crates/networking/p2p/peer_handler.rs` | ETH protocol requests (~670 lines) |
| `test/tests/p2p/snap_server_tests.rs` | Snap server tests (in root test crate) |

### Implementation Order (Dependencies)

```
Phase 1.1-1.2 (snap module)
    ↓
Phase 1.3 (constants) ──→ Phase 2.1-2.3 (rlpx reorganization)
    ↓
Phase 1.4-1.5 (tests, imports)
    ↓
Phase 3.1-3.2 (pending_nodes)
    ↓
Phase 3.3-3.5 (healing unification)
    ↓
Phase 4.1-4.3 (sync split)
    ↓
Phase 4.4-4.5 (snap client extraction)
    ↓
Phase 5.1-5.3 (error consolidation)
```

### Risk Mitigation

| Phase | Risk | Mitigation |
|-------|------|------------|
| 1. Foundation | Low | Simple reorganization, APIs unchanged |
| 2. Protocol | Medium | Extensive hive testing |
| 3. Healing | Medium-High | Incremental migration, keep old files until verified |
| 4. Sync orchestration | High | Feature flags, integration tests, gradual extraction |
| 5. Error handling | Medium | Keep old errors, wrap in new type initially |

### Decisions Made
- **No RLPxMessage macro**: Implementations vary too much (e.g., `GetStorageRanges` special hash handling)
- **Backward-compatible re-exports**: Constants re-exported from `peer_handler.rs` to avoid breaking changes
- **Incremental approach**: Each phase builds on previous, allowing verification at each step

### Key Considerations
- The `accounts_by_root_hash` structure in sync is unbounded - consider adding limits
- Tests should cover edge cases for hash boundary handling in `GetStorageRanges`
- Healing processes share similar patterns but have distinct algorithms - trait unification should preserve this

</details>

### CODE REVIEW

File: crates/networking/p2p/sync/healing/mod.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+pub mod state;
+pub mod storage;
+mod types;
+
+pub use state::heal_state_trie_wrap;
+pub use storage::heal_storage_trie;
+
+// Re-export shared types for external use
+#[allow(unused_imports)]
```
[the remark this answers is not in the document]
Comment (pablodeymo) [reply crates/networking/p2p/sync/healing/mod.rs:14]: I think it's better to leave this annotation.

**Comment by ilitteri** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

**Review summary:**
- CLAUDE.md compliance: The refactoring follows good practices - meaningful names, proper edge case handling, no scope creep
- Bug scan: No obvious logic errors; the PR actually *fixes* potential panics (empty vector indexing, zero chunk_size)
- Git history: All critical bug fixes from previous commits were preserved
- Previous PR comments: No applicable feedback patterns violated
- Code comments: All SAFETY and NOTE comments remain valid after refactoring

The 5-phase reorganization appears well-executed with improved error handling and code organization.

File: crates/networking/p2p/snap/error.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+/// Converts a tokio task JoinError into SnapError
+impl From<tokio::task::JoinError> for SnapError {
+    fn from(err: tokio::task::JoinError) -> Self {
+        SnapError::TaskPanic(err.to_string())
+    }
+}
+
+/// Error that occurs when dumping snapshots to disk
+pub struct DumpError {
```
Comment (ilitteri) [inline crates/networking/p2p/snap/error.rs:134]: This should derive `thiserror::Error`

Comment (pablodeymo) [reply crates/networking/p2p/snap/error.rs:134]: done

File: crates/networking/p2p/sync/healing/types.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+use std::collections::HashMap;
+
+use ethrex_trie::{Nibbles, Node};
+
+/// Entry in the healing queue tracking nodes waiting for children
+#[derive(Debug, Clone)]
+pub struct HealingQueueEntry {
+    pub node: Node,
+    pub missing_children_count: u64,
```
[the remark this answers is not in the document]
Comment (pablodeymo) [reply crates/networking/p2p/sync/healing/types.rs]: Already fixed in commit f6807779d. Both `HealingQueueEntry` and `StorageHealingQueueEntry` now use `usize` for `missing_children_count`.

File: crates/networking/p2p/snap/client.rs [hunk 4]
```diff
@@ -0,0 +1,1416 @@
+//! Snap sync client - methods for requesting snap protocol data from peers
+//!
+//! This module contains all the client-side snap protocol request methods
+//! implemented as extension methods on PeerHandler.
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: I don't think it should be exenstion methods on peer handler. I think it would be better if implemented as just some functions we use a snap sync downloader.

File: crates/networking/p2p/snap/client.rs [hunk 5]
```diff
@@ -0,0 +1,1416 @@
...
+                    accounts.len(),
+                    peer_id
+                );
+                all_account_hashes.extend(accounts.iter().map(|unit| unit.hash));
+                all_accounts_state.extend(accounts.iter().map(|unit| unit.account));
+            }
+
+            let Some((peer_id, connection)) = self
+                .peer_table
+                .get_best_peer(&SUPPORTED_ETH_CAPABILITIES)
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: Wrong capabilities, it's `SUPPORTED_SNAP_CAPABILITIES`

File: crates/networking/p2p/snap/client.rs [hunk 6]
```diff
@@ -0,0 +1,1416 @@
...
+    /// # Returns
+    ///
+    /// The account range or `None` if:
+    ///
+    /// - There are no available peers (the node just started up or was rejected by all other nodes)
+    /// - No peer returned a valid response in the given time and retry limits
+    pub async fn request_account_range(
+        &mut self,
+        start: H256,
+        limit: H256,
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: Only ever gets called with 

`H256::zero(), H256::repeat_byte(0xff)` as the two parameters. May not work with an arbitrary limit.

File: crates/networking/p2p/snap/client.rs [hunk 7]
```diff
@@ -0,0 +1,1416 @@
...
+        block_sync_state: &mut SnapBlockSyncState,
+    ) -> Result<(), SnapError> {
+        METRICS
+            .current_step
+            .set(CurrentStepValue::RequestingAccountRanges);
+        // 1) split the range in chunks of same length
+        let start_u256 = U256::from_big_endian(&start.0);
+        let limit_u256 = U256::from_big_endian(&limit.0);
+
+        let chunk_count = 800;
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: Should be in a well defined `CONST`.

File: crates/networking/p2p/sync/healing/state.rs [hunk 8]
```diff
@@ -416,9 +405,9 @@ pub fn node_missing_children(
     node: &Node,
     path: &Nibbles,
     trie_state: &dyn TrieDB,
-) -> Result<(u64, Vec<RequestMetadata>), TrieError> {
+) -> Result<(usize, Vec<RequestMetadata>), TrieError> {
     let mut paths: Vec<RequestMetadata> = Vec::new();
-    let mut missing_children_count = 0_u64;
```
Comment (fedacking) [inline crates/networking/p2p/sync/healing/state.rs:421]: Is this better as an usize instead of u64?

File: crates/networking/p2p/snap/client.rs [hunk 9]
```diff
@@ -0,0 +1,1416 @@
...
+                                let (_, intervals) = account_storage_roots
+                                    .accounts_with_storage_root
+                                    .get_mut(&first_acc_hash)
+                                    .ok_or(SnapError::InternalError("Tried to get the old download intervals for an account but did not find them".to_owned()))?;
+
+                                for i in 0..chunk_count {
+                                    let start_hash_u256 = start_hash_u256 + chunk_size * i;
+                                    let start_hash = H256::from_uint(&start_hash_u256);
+                                    let end_hash = if i == chunk_count - 1 {
+                                        H256::repeat_byte(0xff)
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: We should replace `H256::repeat_byte(0xff)` with the constant HASH_MAX

File: crates/networking/p2p/snap/constants.rs
Comment (fedacking) [inline crates/networking/p2p/snap/constants.rs:1]: We should make an issue to test different values for these constants in a challenging scenario for snapsync.

File: crates/networking/p2p/snap/error.rs [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        SnapError::TaskPanic(err.to_string())
+    }
+}
+
+/// Error that occurs when dumping snapshots to disk
+#[derive(thiserror::Error)]
+#[error("Failed to dump snapshot to {}: {:?}", path.display(), error)]
+pub struct DumpError {
+    pub path: PathBuf,
+    pub contents: Vec<u8>,
```
Comment (fedacking) [inline crates/networking/p2p/snap/error.rs]: This error has the contents because the old logic had a retry mechanism. This mechanism doesn't exist anymore. We shouldn't keep the content in memory.

File: crates/networking/p2p/snap/error.rs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    pub error: ErrorKind,
+}
+
+impl core::fmt::Debug for DumpError {
```
Comment (fedacking) [inline crates/networking/p2p/snap/error.rs]: Once contents are removed, just use the default debug print.

File: crates/networking/p2p/snap/server.rs [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+            id: request.id,
+            slots,
+            proof,
+        })
+    })
+    .await
+    .map_err(|e| SnapError::TaskPanic(e.to_string()))?
+}
+
+pub fn process_byte_codes_request(
```
Comment (fedacking) [inline crates/networking/p2p/snap/server.rs]: This function doesn't have a spawn blocking as compared to the `process_account_range_request`, `process_storage_ranges_request` and `process_trie_nodes_request`

File: crates/networking/p2p/sync/healing/state.rs [hunk 13]
```diff
@@ -112,11 +102,10 @@ async fn heal_state_trie(
     let mut db_joinset = tokio::task::JoinSet::new();
 
     // channel to send the tasks to the peers
-    let (task_sender, mut task_receiver) = tokio::sync::mpsc::channel::<(
-        H256,
-        Result<Vec<Node>, RequestStateTrieNodesError>,
-        Vec<RequestMetadata>,
-    )>(1000);
+    let (task_sender, mut task_receiver) =
+        tokio::sync::mpsc::channel::<(H256, Result<Vec<Node>, SnapError>, Vec<RequestMetadata>)>(
```
Comment (fedacking) [inline crates/networking/p2p/sync/healing/state.rs:107]: `(H256, Result<Vec<Node>, SnapError>, Vec<RequestMetadata>)` is a complex return type, it should probably be a struct to name and comment what they are.

File: crates/networking/p2p/sync/healing/state.rs [hunk 14]
```diff
@@ -365,16 +354,16 @@ fn heal_state_batch(
...
-                membatch,
+                healing_queue,
                 nodes_to_write,
             );
         } else {
-            let entry = MembatchEntryValue {
+            let entry = HealingQueueEntry {
                 node: node.clone(),
-                children_not_in_storage_count: missing_children_count,
+                missing_children_count,
```
Comment (fedacking) [inline crates/networking/p2p/sync/healing/state.rs]: Opinion: Missing children isn't a good name, I prefer pending children.

File: crates/networking/p2p/sync/healing/mod.rs
Comment (fedacking) [inline crates/networking/p2p/sync/healing/mod.rs:1]: Ideally, there should be a lot more code that is shared between the state and storage healing algorithm, as they're identical in it's conceptual logic. The main differences come from the path being two sets of nibbles for storage and the leafs being accounts vs u256 values.

File: crates/networking/p2p/snap/client.rs [hunk 15]
```diff
@@ -0,0 +1,1416 @@
...
+
+        let mut completed_tasks = 0;
+        let mut chunk_file = 0;
+        let mut last_update: SystemTime = SystemTime::now();
+        let mut write_set = tokio::task::JoinSet::new();
+
+        let mut logged_no_free_peers_count = 0;
+
+        loop {
+            if all_accounts_state.len() * size_of::<AccountState>() >= RANGE_FILE_CHUNK_SIZE {
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: This code path should be shared with the write to disk of line 287. This duplication of code also exists for downloading storages.

File: crates/networking/p2p/snap/client.rs [hunk 16]
```diff
@@ -0,0 +1,1416 @@
...
+                    })?;
+                }
+
+                let account_state_snapshots_dir_cloned = account_state_snapshots_dir.to_path_buf();
+                write_set.spawn(async move {
+                    let path = get_account_state_snapshot_file(
+                        &account_state_snapshots_dir_cloned,
+                        chunk_file,
+                    );
+                    // TODO: check the error type and handle it properly
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: The error is now just printed to window, and the execution is stopped. This seems good enough as long as we don't have user reports of problems with this due to some spurious filesystem behaviour.

File: crates/networking/p2p/snap/client.rs [hunk 17]
```diff
@@ -0,0 +1,1416 @@
...
+                })? {
+                    std::fs::create_dir_all(account_state_snapshots_dir).map_err(|_| {
+                        SnapError::SnapshotDir(
+                            "Failed to create state snapshots directory".to_string(),
+                        )
+                    })?;
+                }
+
+                let account_state_snapshots_dir_cloned = account_state_snapshots_dir.to_path_buf();
+                write_set.spawn(async move {
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: We shouldn't have multiple writing tasks in queue, the write_set should check that there isn't any task in waiting.

File: crates/networking/p2p/snap/client.rs [hunk 18]
```diff
@@ -0,0 +1,1416 @@
...
+        last_task.1 = limit;
+
+        // 2) request the chunks from peers
+
+        let mut downloaded_count = 0_u64;
+        let mut all_account_hashes = Vec::new();
+        let mut all_accounts_state = Vec::new();
+
+        // channel to send the tasks to the peers
+        let (task_sender, mut task_receiver) =
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: I think for a structured concurrency model like this one, it's better to have the workers be single shot tasks that are tracked through a joinset, instead of using a channel to respond. We would avoid any scenario a worker dies without being able to send a message through the channel.

File: crates/networking/p2p/snap/client.rs [hunk 19]
```diff
@@ -0,0 +1,1416 @@
...
+            tokio::sync::mpsc::channel::<(Vec<AccountRangeUnit>, H256, Option<(H256, H256)>)>(1000);
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: We should have a struct that represents the return of the worker in a better more structured result.

File: crates/networking/p2p/sync.rs
Comment (fedacking) [inline crates/networking/p2p/sync.rs:167]: This is the critical struct for downloading storages. It's the single most complicated step, and as such simplifying this struct while preserving the key performance characteristics is needed to improve snapsync. I strongly recommend a data oriented structure of arrays approach.

File: crates/networking/p2p/snap/client.rs [hunk 20]
```diff
@@ -0,0 +1,1416 @@
...
+                        .entry(root)
+                        .or_default()
+                        .push(*account);
+                }
+            }
+        }
+        let mut accounts_by_root_hash = Vec::from_iter(accounts_by_root_hash);
+        // TODO: Turn this into a stable sort for binary search.
+        accounts_by_root_hash.sort_unstable_by_key(|(_, accounts)| !accounts.len());
+        let chunk_size = 300;
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: Magic number: should be in constants.rs

File: crates/networking/p2p/snap/client.rs [hunk 21]
```diff
@@ -0,0 +1,1416 @@
...
+                    accounts_by_root_hash
+                        .entry(*root)
+                        .or_default()
+                        .push(*account);
+                }
+                None => {
+                    let root = store
+                        .get_account_state_by_acc_hash(pivot_header.hash(), *account)
+                        .expect("Failed to get account in state trie")
+                        .expect("Could not find account that should have been downloaded or healed")
```
Comment (iovoid) [inline crates/networking/p2p/snap/client.rs]: Can't we propagate the error?

File: crates/networking/p2p/snap/server.rs [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+#[inline]
+pub(crate) fn proof_to_encodable(proof: Vec<Vec<u8>>) -> Vec<Bytes> {
+    proof.into_iter().map(Bytes::from).collect()
+}
+
+// Helper method to obtain proof from RLP-encodable format
+#[inline]
+pub(crate) fn encodable_to_proof(proof: &[Bytes]) -> Vec<Vec<u8>> {
+    proof.iter().map(|bytes| bytes.to_vec()).collect()
+}
```
Comment (iovoid) [inline crates/networking/p2p/snap/server.rs]: This is more of an utility, it should be elsewhere. encodable_to_proof is used both for the client and server

File: crates/networking/p2p/snap/client.rs [hunk 23]
```diff
@@ -0,0 +1,1416 @@
...
+    account_storages: Vec<Vec<(H256, U256)>>,
+    peer_id: H256,
+    remaining_start: usize,
+    remaining_end: usize,
+    remaining_hash_range: (H256, Option<H256>),
+}
+
+#[derive(Debug)]
+struct StorageTask {
+    start_index: usize,
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs:76]: Sending the tasks by index is error prone as all task now have this global state by the index. If the array of hashes changes at any point, these tasks become invalid. We should send the hashes as tasks in my opinion. Changing this would substantially change the `request_storage_ranges` function and `AccountStorageRoots` struct.

File: crates/networking/p2p/snap/client.rs [hunk 24]
```diff
@@ -0,0 +1,1416 @@
...
+    pub source: SnapError,
+}
+
+#[derive(Clone)]
+struct StorageTaskResult {
+    start_index: usize,
+    account_storages: Vec<Vec<(H256, U256)>>,
+    peer_id: H256,
+    remaining_start: usize,
+    remaining_end: usize,
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs:70]: `remaining_start` `remaining_end` and `remaining_hash_range` are obtuse without an explanation.

File: crates/networking/p2p/snap/client.rs [hunk 25]
```diff
@@ -0,0 +1,1416 @@
...
+            tasks_queue_not_started.push_back(StorageTask {
+                start_index: chunk_start,
+                end_index: chunk_end,
+                start_hash: H256::zero(),
+                end_hash: None,
+            });
+        }
+
+        // channel to send the tasks to the peers
+        let (task_sender, mut task_receiver) =
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: Same issue, we should handle all task with a joinset.

File: crates/networking/p2p/snap/client.rs [hunk 26]
```diff
@@ -0,0 +1,1416 @@
...
+        mut chunk_index: u64,
+        pivot_header: &mut BlockHeader,
+        store: Store,
+    ) -> Result<u64, SnapError> {
+        METRICS
+            .current_step
+            .set(CurrentStepValue::RequestingStorageRanges);
+        debug!("Starting request_storage_ranges function");
+        // 1) split the range in chunks of same length
+        let mut accounts_by_root_hash: BTreeMap<_, Vec<_>> = BTreeMap::new();
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: Key insight: the `accounts_by_root_hash` is the accounts organized by the hash of the storage root. This is key because each storage root is downloaded only once. We still need to know the path to that account.

File: crates/networking/p2p/sync.rs
Comment (fedacking) [inline crates/networking/p2p/sync.rs:170]: The `(Option<H256>, Vec<(H256, H256)>)` is a complex type. Here's what it means, and it should be made more explicit. `Option<H256>` is the storage root of the account. If the root is `None` it means that we healed the account and we aren't sure if that is the storage root anymore. This behaviour was needed in the hash based trie, it may not be necessary in path based trie. `Vec<(H256, H256)>` is the intervals that we still need to download from the storage trie. When we reach a big account, we chunk it for download the storage. During the download we may need to stop request_strorage_request and continue later. As such, we need to know which intervals are still missing.

File: crates/networking/p2p/snap/client.rs [hunk 27]
```diff
@@ -0,0 +1,1416 @@
...
+                                .ok_or(SnapError::InternalError("Tried to get the old download intervals for an account but did not find them".to_owned()))?;
+
+                            if old_intervals.is_empty() {
+                                accounts_done.insert(*account, vec![]);
+                            }
+                        }
+                    }
+                }
+
+                if remaining_start < remaining_end {
```
Comment (fedacking) [inline crates/networking/p2p/snap/client.rs]: The logic with remaining_start and remaining_end logic is a complex state machine. We should make it more explicit so it's easier to understand, ideally by changing the StorageResult struct. The state machine can be viewed in the documentation at the document `docs/l1/fundamentals/snap_sync/Flow-BigAccountLogic.png`

File: crates/networking/p2p/snap/client.rs (hunk 21 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/client.rs]: Done. Replaced both `.expect()` calls with `?` propagation — the `StoreError` converts to `SnapError` directly, and the `None` case now returns `SnapError::InternalError`.

File: crates/networking/p2p/snap/server.rs (hunk 22 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/server.rs]: Agreed. Moved both `encodable_to_proof` and `proof_to_encodable` to `snap/mod.rs` since they're shared between client and server.

File: crates/networking/p2p/snap/client.rs (hunk 5 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/client.rs]: Done. Changed all three occurrences in `client.rs` to use `SUPPORTED_SNAP_CAPABILITIES`.

File: crates/networking/p2p/snap/client.rs (hunk 7 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/client.rs]: Done. Extracted to `ACCOUNT_RANGE_CHUNK_COUNT` in `constants.rs`.

File: crates/networking/p2p/snap/client.rs (hunk 20 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/client.rs]: Done. Replaced with `STORAGE_BATCH_SIZE` which was already defined in `constants.rs`.

File: crates/networking/p2p/snap/client.rs (hunk 9 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/client.rs]: Done. Both occurrences replaced with `HASH_MAX`.

File: crates/networking/p2p/snap/error.rs (hunk 10 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/error.rs]: Done. Removed the `contents` field from `DumpError` and updated all construction sites.

File: crates/networking/p2p/snap/error.rs (hunk 11 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/error.rs]: Done. Using `derive(Debug)` now.

File: crates/networking/p2p/sync/healing/state.rs (hunk 14 above)
Comment (pablodeymo) [reply crates/networking/p2p/sync/healing/state.rs]: Done. Renamed to `pending_children` / `pending_children_count` across both `state.rs` and `storage.rs`.

File: crates/networking/p2p/snap/server.rs (hunk 12 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/server.rs]: Done. Wrapped `process_byte_codes_request` in `spawn_blocking` for consistency with the other handlers.

File: crates/networking/p2p/sync/healing/state.rs (hunk 8 above)
Comment (pablodeymo) [reply crates/networking/p2p/sync/healing/state.rs:421]: It was already `usize` — this was fixed in a previous commit (f6807779d).

File: crates/networking/p2p/snap/constants.rs
Comment (pablodeymo) [reply crates/networking/p2p/snap/constants.rs:1]: Good idea. Will create an issue for this.

File: crates/networking/p2p/snap/client.rs (hunk 4 above)
Comment (pablodeymo) [reply crates/networking/p2p/snap/client.rs]: Done. Converted all 5 methods out of `impl PeerHandler` into standalone `pub async fn` functions. The three that used `self.peer_table` (`request_account_range`, `request_bytecodes`, `request_storage_ranges`) now take `peers: &mut PeerHandler` as a parameter. The two that were already static (`request_state_trienodes`, `request_storage_trienodes`) just moved out of the impl block unchanged.

File: crates/networking/p2p/sync.rs [hunk 28]
```diff
@@ -1,21 +1,21 @@
...
 
-use crate::peer_handler::{BlockRequestOrder, PeerHandlerError, SNAP_LIMIT};
+use crate::peer_handler::{BlockRequestOrder, PeerHandler, PeerHandlerError};
 use crate::peer_table::PeerTableError;
 use crate::rlpx::p2p::SUPPORTED_ETH_CAPABILITIES;
+use crate::snap::constants::{
+    BYTECODE_CHUNK_SIZE, EXECUTE_BATCH_SIZE_DEFAULT, MAX_BLOCK_BODIES_TO_REQUEST,
+    MAX_HEADER_FETCH_ATTEMPTS, MIN_FULL_BLOCKS, MISSING_SLOTS_PERCENTAGE, SECONDS_PER_BLOCK,
+    SNAP_LIMIT,
+};
```
[the remark this answers is not in the document]
Comment (pablodeymo) [reply crates/networking/p2p/sync.rs]: This was addressed — the constants were moved to `snap::constants` and the imports compile correctly now.
